File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/UpdatePriceService.java
package com.brentpanther.bitcoinwidget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;


import androidx.annotation.NonNull;
import androidx.core.app.JobIntentService;

import static android.app.PendingIntent.FLAG_CANCEL_CURRENT;


public class UpdatePriceService extends JobIntentService {

    private static final int JOB_ID = 3542;
    public static final String EXTRA_MANUAL_REFRESH = "manualRefresh";

    static void enqueueWork(Context context, Intent work) {
        DataMigration.migrate(context);
        enqueueWork(context, UpdatePriceService.class, JOB_ID, work);
    }

    @Override
    protected void onHandleWork(@NonNull Intent intent) {
        int appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, 0);
        boolean manualRefresh = intent.getBooleanExtra(EXTRA_MANUAL_REFRESH, false);
        Context context = getApplicationContext();
        Prefs prefs = new Prefs(appWidgetId);
        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
        int layout = prefs.getThemeLayout();
        RemoteViews views = new RemoteViews(context.getPackageName(), layout);
        String currencyCode = prefs.getExchangeCurrencyName();
        if (currencyCode == null) return;
        try {
            Exchange exchange = prefs.getExchange();
            String amount = exchange.getValue(prefs.getExchangeCoinName(), currencyCode);
            WidgetViews.setText(context, views, amount, prefs);
            prefs.setLastUpdate();
            setOldValue(prefs, views, context);
        } catch (IllegalArgumentException e) {
            float textSize = WidgetViews.putValue(context, views, context.getString(R.string.value_exchange_removed), prefs);
            appWidgetManager.updateAppWidget(appWidgetId, views);
            return;
        } catch (Exception e) {
            setOldValue(prefs, views, context);
        }
        Intent priceUpdate = new Intent(context, PriceBroadcastReceiver.class);
        priceUpdate.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        priceUpdate.putExtra(EXTRA_MANUAL_REFRESH, true);
        PendingIntent pendingPriceUpdate = PendingIntent.getBroadcast(context, appWidgetId, priceUpdate, FLAG_CANCEL_CURRENT);
        views.setOnClickPendingIntent(R.id.parent, pendingPriceUpdate);
        appWidgetManager.updateAppWidget(appWidgetId, views);
    }

    private static void setOldValue(Prefs prefs, RemoteViews views, Context context) {
        long lastUpdate = prefs.getLastUpdate();
        int interval = prefs.getInterval();
        //if its been "a while" since the last successful update, gray out the icon.
        boolean isOld = ((System.currentTimeMillis() - lastUpdate) > 1000 * 90 * interval);
        WidgetViews.setLastText(context, views, prefs);
        WidgetViews.setOld(context, views, isOld, prefs);
    }

}

File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/UpdatePriceService.kt
package com.brentpanther.bitcoinwidget


import android.appwidget.AppWidgetManager
import android.content.Intent
import android.widget.RemoteViews
import androidx.core.app.JobIntentService
import com.brentpanther.bitcoinwidget.db.Widget
import com.brentpanther.bitcoinwidget.db.WidgetDatabase
import com.brentpanther.bitcoinwidget.db.WidgetSettings
import com.brentpanther.bitcoinwidget.exchange.ExchangeHelper
import com.brentpanther.bitcoinwidget.receiver.PriceBroadcastReceiver
import com.brentpanther.bitcoinwidget.ui.preview.RemoteWidgetPreview
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


class UpdatePriceService : JobIntentService() {

    override fun onHandleWork(intent: Intent) {
        val widgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, 0)
        val dao = WidgetDatabase.getInstance(applicationContext).widgetDao()
        val widget = dao.getByWidgetId(widgetId) ?: return
        val views = RemoteViews(applicationContext.packageName, widget.theme.layout)

        val widgetSettings = WidgetSettings(widget, dao.configWithSizes())
        val widgetViews = WidgetViews(applicationContext, RemoteWidgetPreview(views), widgetSettings)
        when (val amount = updateValue(widget)) {
            INVALID -> widgetViews.putValue(applicationContext.getString(R.string.value_exchange_removed))
            else -> {
                widgetViews.setText(amount, true)
                CoroutineScope(Dispatchers.IO).launch {
                    dao.update(widget)
                }
            }
        }
        val appWidgetManager = AppWidgetManager.getInstance(applicationContext)
        PriceBroadcastReceiver.setOnClick(applicationContext, widgetId, views)
        appWidgetManager.updateAppWidget(widgetId, views)
    }

    companion object {

        const val INVALID = "invalid"
        private val TAG = UpdatePriceService::class.java.simpleName

        internal fun updateValue(widget: Widget): String? {
            return try {
                val currencyCode = widget.currencyCustomName ?: widget.currency
                val exchange = widget.exchange
                val amount = exchange.getValue(widget.coinCustomName ?: widget.coin.name, currencyCode) ?: return null
                widget.lastUpdated = System.currentTimeMillis()
                amount
            } catch (e: IllegalArgumentException) {
                INVALID
            } catch (ignored: Exception) {
                ExchangeHelper.connectionPool.evictAll()
                null
            }
        }
    }

}
